### DISCUSSION TITLE
Support for Xtext enum

### DISCUSSION THREAD

**Discussion by brsanthu** [eclipse-langium/langium#1475] [Q&A] [answered]
We had a language in Xtext. We are migrating into Langium.

Xtext had concept of enum values like this.

```
enum ForLiterals:
    update
    | view
    | reference
;
```

I haven't seen any mention of this in Langium grammar. 

1. Is there an equivalent concept?
2. Or support for it is not there yet and will be added?

**Comment by Yokozuna59** [comment, marked as answer]
I'd recommend you read this article: <https://www.typefox.io/blog/xtext-to-langium/> and this discussion: https://github.com/eclipse-langium/langium/discussions/521.

But answering your question using the article approach, then it would be something like this:

Xtext:

```xtext
enum ForLiterals: update | view | reference;
```

Langium:

```langium
type ForLiterals = 'update' | 'view' | 'reference';
ForLiterals returns ForLiterals:
	ForLiterals_Update | ForLiterals_View | ForLiterals_Reference
;
ForLiterals_Update returns string: 'update';
ForLiterals_View returns string: 'view';
ForLiterals_Reference returns string: 'reference';
```

**Reply by msujew** [reply]
I think you can shorten this to just:
```
ForLiterals returns string:
    'update'
    | 'view'
    | 'reference'
;
```

**Comment by brsanthu** [comment]
Thank you for that great reference. We could convert our old xtext based grammar into Langium. There were couple of issues. The ML and SL comments were not working when we migrated. We had to redo that part.

Other than that we haven't seen other issues. Will keep an eye and post about any issues here.

**Comment by ebousse** [comment]
Hi, sorry to bother this old thread but I am trying to do something similar, without success.

I have the following grammar:

```langium
interface Root {
    color: Color
}

type Color = 'red' | 'blue' | 'green';


entry RootRule returns Root:
    'color' color=ColorRule
;

ColorRule returns Color:
	 'red'
    | 'blue'
    | 'green'
;
```

And I get this error:

```
The assigned type 'ColorRule' is not compatible with the declared property 'color' of type 'Color'.
```

Screenshot:
<img width="884" height="314" alt="image" src="https://github.com/user-attachments/assets/843e42ca-d191-418c-be63-cbc205a0fa61" />

What am I doing wrong? :(

**Reply by msujew** [reply]
Hey @ebousse, there seems to be something weird going on with the validator. If you rename the `ColorRule` simply to `Color` (the same as the type you're generating), it seems to work. It looks like some bug has managed to sneak into the type system.

**Reply by ebousse** [reply]
Oh I see, then I'll open a bug report. Thanks!

**Reply by ebousse** [reply]
I opened an issue here: https://github.com/eclipse-langium/langium/issues/2074
